feat: create max-nested-describe rule

dominictwlee opened the pull request:
Resolves #362

The idea is the same as the above issue. I changed the rule name and options to something more consistent with the `max-nested-callbacks` rule.

G-Rath commented on src/rules/utils.ts:
```diff
@@ -648,6 +648,16 @@ export type FunctionExpression =
   | TSESTree.ArrowFunctionExpression
   | TSESTree.FunctionExpression;
 
+export const isCallExpression = (
```
We don't need this utility function, as you can just do `node?.type === AST_NODE_TYPES.CallExpression`

G-Rath commented on src/rules/max-nested-describe.ts:
```diff
@@ -0,0 +1,79 @@
+    schema: [
+      {
+        oneOf: [
+          {
```
Let's keep things simple and just support an object

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,135 @@
+  },
+});
+
+ruleTester.run('max-nested-describe', rule, {
```
Could you add a test for `{ max: 0 }`

G-Rath replied:
Could you chuck in a few tests for prefixed & modified describes? (i.e `fdescribe`, `describe.skip` - feel free to just chuck these into some of the existing tests), and one of each version of `describe.each`.

dominictwlee commented:
Thanks for the review @G-Rath, I've made the changes according to your comments.

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+  "jest/max-nested-describe": [
+    "error",
+    {
+      "max": 3
```
```suggestion
      "max": 2
```

nit: I think it's better if we use the default value here.

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+describe('foo', () => {
+  describe('bar', () => {
+    it('should get something', () => {
+      expect(getSomething().toBe('Something'));
```
```suggestion
      expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+
+  describe('qux', () => {
+    it('should get something', () => {
+      expect(getSomething().toBe('Something'));
```
```suggestion
      expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+
+describe('foo2', function () {
+  it('should get something', () => {
+    expect(getSomething().toBe('Something'));
```
```suggestion
    expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+  describe('bar2', function () {
+    describe('baz2', function () {
+      it('should get something', () => {
+        expect(getSomething().toBe('Something'));
```
```suggestion
        expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+  describe('bar', () => {
+    describe('baz', () => {
+      it('should get something', () => {
+        expect(getSomething().toBe('Something'));
```
```suggestion
        expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+  describe('bar', () => {
+    describe('baz', () => {
+      it('should get something', () => {
+        expect(getSomething().toBe('Something'))
```
```suggestion
        expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+  describe('bar2', function() {
+    describe('baz2', function() {
+      it('should get something', function() {
+        expect(getSomething().toBe('Something'))
```
```suggestion
        expect(getSomething()).toBe('Something');
```

G-Rath commented on docs/rules/max-nested-describe.md:
```diff
@@ -0,0 +1,111 @@
+
+    describe('qux2', function() {
+      it('should get something', function() {
+        expect(getSomething().toBe('Something'))
```
```suggestion
        expect(getSomething()).toBe('Something');
```

G-Rath reviewed (changes requested):
Cheers, this is looking good! Just got some documentation fixes, and if you could add a few modifier & `.each` tests that'd be great!

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+            describe('baz', function () {
+              describe('qux', function () {
+                it('should get something', () => {
+                  expect(getSomething().toBe('Something'))
```
```suggestion
                  expect(getSomething()).toBe('Something');
```

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+          describe('bar', () => {
+            describe('baz', () => {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
```suggestion
                expect(getSomething()).toBe('Something');
```

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+
+            describe('qux', function () {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
```suggestion
                expect(getSomething()).toBe('Something');
```

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+          describe.only('bar', () => {
+            describe.skip('baz', () => {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
```suggestion
                expect(getSomething()).toBe('Something');
```

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+
+            describe('baz', () => {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
```suggestion
                expect(getSomething()).toBe('Something');
```

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+
+        xdescribe('qux', () => {
+          it('should get something', () => {
+            expect(getSomething().toBe('Something'))
```
```suggestion
            expect(getSomething()).toBe('Something');
```

G-Rath commented on src/rules/__tests__/max-nested-describe.test.ts:
```diff
@@ -0,0 +1,197 @@
+      code: dedent`
+        describe('qux', () => {
+          it('should get something', () => {
+            expect(getSomething().toBe('Something'))
```
```suggestion
            expect(getSomething()).toBe('Something');
```

dominictwlee commented:
My bad with the brackets! been a long day my eyes must be going fuzzy.

G-Rath commented on src/rules/max-nested-describe.ts:
```diff
@@ -0,0 +1,80 @@
+    },
+    messages: {
+      exceededMaxDepth:
+        'Too many nested describe calls ({{depth}}). Maximum allowed is {{max}}.',
```
```suggestion
        'Too many nested describe calls ({{ depth }}). Maximum allowed is {{ max }}.',
```

nit: I typically put spaces around the names (and so most if not all of our rules are like this) - we don't have a lint rule for enforcing this, but would be nice to be consistent.

G-Rath reviewed (approved):
Awesome! No worries about the brackets - thanks for putting up with all my nitty requests 😄

G-Rath commented:
I'm now wondering if `2` is a good max, given that `max-nested-callbacks` has a max of `10`.

Thinking about the tests I've written, I easily have more than two nested `describe`s, and think you would easily hit that with class/method tests:

```
describe('MyClass', () => {
  describe('#doSomethingWithDate', () => {
    describe('when an error happens', () => {
       ...
    });

    describe('when given a number', () => {
      it('parses it as a date', () => {
         ...
      });
    });

    describe('when the date is in the future', () => {
       ...
    });
  });
});
```

I'm not overly fussed about having the perfect default, but wouldn't mind hearing if @SimenB has any opinions on what a good nested level might be.

dominictwlee commented:
> I'm now wondering if `2` is a good default max, given that `max-nested-callbacks` has a max of `10`.
> 
> Thinking about the tests I've written, I easily have more than two nested `describe`s, and think you would easily hit that with class/method tests:
> 
> ```
> describe('MyClass', () => {
>   describe('#doSomethingWithDate', () => {
>     describe('when an error happens', () => {
>        ...
>     });
> 
>     describe('when given a number', () => {
>       it('parses it as a date', () => {
>          ...
>       });
>     });
> 
>     describe('when the date is in the future', () => {
>        ...
>     });
>   });
> });
> ```
> 
> I'm not overly fussed about having the perfect default, but wouldn't mind hearing if @SimenB has any opinions on what a good nested level might be.

Yeah it's pretty subjective so probably depends on the respective individuals/teams.Though I do think 10 is probably a bit much for nested describes.

dominictwlee commented:
@SimenB @G-Rath Hey just following up on this. Any thoughts on the dafaults?

G-Rath commented:
@dominictwlee sorry I've been very busy recently, both working on some core changes to `eslint-plugin-jest` (currently, looking at using `settings` for configuring rules) and with other work stuff 😅

Currently I think 5 is probably a good default, as it leaves you with 3 levels of nesting after a class describe + method describe.

dominictwlee commented:
> @dominictwlee sorry I've been very busy recently, both working on some core changes to `eslint-plugin-jest` (currently, looking at using `settings` for configuring rules) and with other work stuff 
> 
> Currently I think 5 is probably a good default, as it leaves you with 3 levels of nesting after a class describe + method describe.

@G-Rath  I've updated the default max to 5, and added a few tests + examples in docs for this.

dominictwlee commented:
@G-Rath Just following up on this periodically, let me know if anything needs to change.

G-Rath commented:
@dominictwlee cheers, lets land it
